Moving the implementation of the constructor from header to the implementation fixes(#7904)

SankarshanMishra opened the pull request:
### **User description**
## Description
Move code from header (.h) to the implementation (.cpp file) fixes (#7904) 

## Checklist
-  Make sure that you are listed in the AUTHORS file
-  Add relevant changes and new features to the CHANGELOG file
-  I have commented my code, particularly in hard-to-understand areas
-  New and existing unit tests pass locally with my changes
- Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)


___

### **PR Type**
Enhancement, Documentation


___

### **Description**
- Moved constructor implementation from header to cpp file.

- Updated AUTHORS file to include contributor's name.

- Updated CHANGELOG to reflect the refactoring changes.


___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>FeatureFinderAlgorithmMetaboIdent.cpp</strong><dd><code>Added constructor implementation in cpp file</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/openms/source/FEATUREFINDER/FeatureFinderAlgorithmMetaboIdent.cpp

<li>Added the constructor definition for <code>FeatureFinderMetaboIdentCompound</code>.<br> <li> Implemented the constructor logic in the cpp file.


</details>


  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7918/files#diff-98a7f4c4849d0988800a5b996351fb897c339faad3d1d050dc70a69a23ca9089">+17/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>FeatureFinderAlgorithmMetaboIdent.h</strong><dd><code>Removed inline constructor from header</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/openms/include/OpenMS/FEATUREFINDER/FeatureFinderAlgorithmMetaboIdent.h

<li>Removed inline constructor implementation from the header file.<br> <li> Declared the constructor for <code>FeatureFinderMetaboIdentCompound</code>.


</details>


  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7918/files#diff-f7c5efac35498cdafa895f6b9ea672f77a6138b0e0a320c2827d626fb125eebd">+1/-10</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>AUTHORS</strong><dd><code>Updated AUTHORS file with contributor name</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

AUTHORS

- Added contributor's name to the AUTHORS file.


</details>


  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7918/files#diff-ab6af77435f58cc0c9d4c31dfe05656e45187cc7c7fc02aada401a7642125463">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>CHANGELOG</strong><dd><code>Updated CHANGELOG for refactoring changes</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

CHANGELOG

- Added entry for moving constructor implementation to cpp file.


</details>


  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7918/files#diff-ecec88c33adb7591ee6aa88e29b62ad52ef443611cba5e0f0ecac9b5725afdba">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Chores**  
  - Updated the contributor list to acknowledge Sankarshan Mishra.

- **Bug Fixes**  
  - Addressed an underlying issue by reorganizing internal code structure for improved system stability.

- **Refactor**  
  - Streamlined internal setup processes to enhance maintainability without affecting user functionality.  
  - Transitioned several member functions in the BaseModel class to declarations only, emphasizing a more abstract interface.  
  - Added a constructor to the FeatureFinderMetaboIdentCompound class for improved instance initialization.
  - Enhanced the BaseModel class with new methods for managing cutoff values and samples.
  - Simplified the EmgFitter1D class by removing unnecessary complexity related to internal structures.  
  - Modified the fit1d method in the EmgFitter1D class to change how position data is accessed.
  - Introduced a new InterpolationModel class with methods for managing interpolation and scaling factors.  
  - Removed comments and streamlined constructor and destructor declarations in the LevMarqFitter1D class.  
  - Added constructors, a destructor, and an assignment operator to the LevMarqFitter1D class for improved resource management.  
  - Removed unnecessary constructors and methods in the MaxLikeliFitter1D class, transitioning to declarations only.  
  - Enhanced the MaxLikeliFitter1D class with constructors and an assignment operator for proper instance management.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

SankarshanMishra committed:
Update FeatureFinderAlgorithmMetaboIdent.h

Removing the implementation from the header file and declaring it.

SankarshanMishra committed:
Update FeatureFinderAlgorithmMetaboIdent.h

SankarshanMishra committed:
Update FeatureFinderAlgorithmMetaboIdent.cpp

Add the implementation to the cpp file.

SankarshanMishra committed:
Update AUTHORS to contain my name

SankarshanMishra committed:
Update CHANGELOG

cbielow commented:
yes, this is going in the right direction.
To make this worthwhile, this would need a lot more of the same (in this PR if possible).

SankarshanMishra committed:
Update BaseModel.h

Update the BaseModel.h file to only declare the functions without implementation.

SankarshanMishra committed:
Update BaseModel.cpp

SankarshanMishra committed:
Update EmgFitter1D.h

SankarshanMishra committed:
Update EmgFitter1D.cpp

move the implementation from the header file to the EmgFitter1D.cpp.

SankarshanMishra committed:
Update InterpolationModel.h

update the header file InterpolationModel.h to only have the method declarations.

SankarshanMishra committed:
Update InterpolationModel.cpp

Move the implementation of each method from the header file (InterpolationModel.h) to the CPP file.

SankarshanMishra committed:
Update LevMarqFitter1D.h

Declare the methods in the class definition without their implementation.

SankarshanMishra committed:
Update LevMarqFitter1D.cpp

Move the method implementations from the header to the CPP file.

SankarshanMishra committed:
Update MaxLikeliFitter1D.h

Update the header file to remove the inline implementations and declare the methods instead;

SankarshanMishra committed:
Update MaxLikeliFitter1D.cpp

move implementation from header to cpp

SankarshanMishra committed:
Merge branch 'OpenMS:develop' into moving-the-implementation-of-the-constructor-from-header-to-the-implementation-fixes(#7904)

cbielow commented on src/openms/include/OpenMS/FEATUREFINDER/EmgFitter1D.h:
```diff
@@ -39,55 +39,17 @@ namespace OpenMS
-    virtual void setInitialParameters_(const RawDataArrayType& set);
-    /// Compute start parameters using method of moments (usually reduces nr. of iterations needed at some
-    /// additional one-time costs
+    void setInitialParameters_(const RawDataArrayType& set);
```
please do not remove the documentation.

cbielow commented on src/openms/include/OpenMS/FEATUREFINDER/EmgFitter1D.h:
```diff
@@ -39,55 +39,17 @@ namespace OpenMS
+    struct Data;
+    class EgmFitterFunctor;
     CoordinateType height_;
-    /// Parameter of emg - peak width
```
also this doc must stay

cbielow commented on src/openms/include/OpenMS/FEATUREFINDER/LevMarqFitter1D.h:
```diff
@@ -19,91 +18,43 @@ namespace Eigen
   using MatrixXd = Matrix<double, -1, -1, 0, -1, -1>;
   using VectorXd = Matrix<double, -1, 1, 0, -1, 1>;
-} // namespace Eigen
+}
```
undo

cbielow commented on src/openms/source/FEATUREFINDER/BaseModel.cpp:
```diff
@@ -9,6 +9,68 @@
+}
+
+// Destructor
+BaseModel::~BaseModel()
```
if you touch the code anyway, think about using ` = default;`. In these cases this is what should be done.

cbielow commented on src/openms/source/FEATUREFINDER/FeatureFinderAlgorithmMetaboIdent.cpp:
```diff
@@ -107,6 +107,23 @@ namespace OpenMS
+                                                                                                         const std::vector<int>& _charges, 
+                                                                                                         const std::vector<double>& _rts, 
+                                                                                                         const std::vector<double>& _rt_ranges, 
+                                                                                                         const std::vector<double>& _iso_distrib) :
```
formatting looks a bit weird. Did you use spaces (no tabs)?

cbielow reviewed (changes requested):
thanks, goes into the correct direction.

SankarshanMishra committed:
Update EmgFitter1D.h

To contain the removed documentation.

SankarshanMishra committed:
Update LevMarqFitter1D.h

SankarshanMishra committed:
Update BaseModel.cpp

simplify constructor, destructor, and copy operations by using = default;  instead of explicitly defining them.

SankarshanMishra committed:
Update FeatureFinderAlgorithmMetaboIdent.cpp

SankarshanMishra committed:
Update src/openms/source/FEATUREFINDER/BaseModel.cpp

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

SankarshanMishra committed:
Update BaseModel.cpp

SankarshanMishra committed:
Merge branch 'develop' into moving-the-implementation-of-the-constructor-from-header-to-the-implementation-fixes(#7904)

SankarshanMishra committed:
Update BaseModel.cpp

timosachsenberg commented:
This is definitely an improvement. I will run CI and see if it breaks anything. Thanks

jpfeuffer commented:
I think it should be best practice if new contributors actually build and test openms on their own!
Just editing in a web editor will likely neither contribute much to the education of the contributor, nor help us improve our onboarding procedure.
With devcontainers there should be ways without installing *anything but docker* and build Openms out of the box.
It would be great @SankarshanMishra if you could setup your own dev environment. This is absolutely mandatory for a successful application in GSOC.

timosachsenberg commented:
yes, please @SankarshanMishra please built and run tests.

SankarshanMishra commented:
@jpfeuffer and @timosachsenberg,  am on it Thank you so much

timosachsenberg commented:
not active... will close. feel free to reopen
